File: backend/app/api/v1/endpoints/notifications.py
"""
Notification API Endpoints
User notification management
"""
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session
from typing import Optional, List
from uuid import UUID
import math

from app.api.deps import get_db, get_current_active_user
from app.models.user import User
from app.models.notification import NotificationType
from app.services.notification_service import NotificationService
from app.schemas.notification import (
    NotificationResponse,
    NotificationListResponse,
    NotificationCountResponse,
    MarkReadRequest,
    MarkAllReadRequest,
    NotificationPreferenceResponse,
    NotificationPreferencesResponse,
    NotificationPreferenceUpdate,
    BulkPreferenceUpdate,
)

router = APIRouter()


# ============================================================================
# NOTIFICATION LISTING
# ============================================================================

@router.get("/", response_model=NotificationListResponse)
async def list_notifications(
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    unread_only: bool = Query(False),
    notification_type: Optional[str] = Query(None),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """
    List notifications for the current user.
    Supports pagination and filtering by read status and type.
    """
    service = NotificationService(db)

    notifications, total = service.list_notifications(
        user_id=current_user.id,
        page=page,
        page_size=page_size,
        unread_only=unread_only,
        notification_type=notification_type,
    )

    unread_count = service.get_unread_count(current_user.id)

    return NotificationListResponse(
        notifications=[
            NotificationResponse.model_validate(n) for n in notifications
        ],
        total=total,
        unread_count=unread_count,
        page=page,
        page_size=page_size,
        total_pages=math.ceil(total / page_size) if total > 0 else 1,
    )


@router.get("/count", response_model=NotificationCountResponse)
async def get_unread_count(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """Get the count of unread notifications"""
    service = NotificationService(db)
    count = service.get_unread_count(current_user.id)
    return NotificationCountResponse(unread_count=count)


@router.get("/types")
async def get_notification_types():
    """Get all available notification types"""
    types = []
    for nt in NotificationType:
        category = nt.value.split(".")[0]
        types.append({
            "type": nt.value,
            "category": category,
            "name": nt.name.replace("_", " ").title(),
        })
    return {"types": types}


@router.get("/{notification_id}", response_model=NotificationResponse)
async def get_notification(
    notification_id: UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """Get a specific notification"""
    service = NotificationService(db)
    notification = service.get_notification(notification_id, current_user.id)

    if not notification:
        raise HTTPException(status_code=404, detail="Notification not found")

    return NotificationResponse.model_validate(notification)


# ============================================================================
# MARK AS READ
# ============================================================================

@router.post("/mark-read")
async def mark_notifications_read(
    data: MarkReadRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """Mark specific notifications as read"""
    service = NotificationService(db)
    count = service.mark_as_read(data.notification_ids, current_user.id)
    return {"marked_read": count}


@router.post("/mark-all-read")
async def mark_all_notifications_read(
    data: Optional[MarkAllReadRequest] = None,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """Mark all notifications as read (optionally filtered by type)"""
    service = NotificationService(db)
    notification_type = data.notification_type if data else None
    count = service.mark_all_as_read(current_user.id, notification_type)
    return {"marked_read": count}


# ============================================================================
# DELETE NOTIFICATIONS
# ============================================================================

@router.delete("/{notification_id}")
async def delete_notification(
    notification_id: UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """Delete a specific notification"""
    service = NotificationService(db)
    deleted = service.delete_notification(notification_id, current_user.id)

    if not deleted:
        raise HTTPException(status_code=404, detail="Notification not found")

    return {"deleted": True}


@router.delete("/")
async def delete_all_read_notifications(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """Delete all read notifications"""
    service = NotificationService(db)
    count = service.delete_all_read(current_user.id)
    return {"deleted": count}


# ============================================================================
# NOTIFICATION PREFERENCES
# ============================================================================

@router.get("/preferences/all", response_model=NotificationPreferencesResponse)
async def get_notification_preferences(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """Get all notification preferences for the current user"""
    service = NotificationService(db)
    preferences = service.get_all_preferences(current_user.id)

    return NotificationPreferencesResponse(
        preferences=[
            NotificationPreferenceResponse.model_validate(p) for p in preferences
        ]
    )


@router.put("/preferences")
async def update_notification_preference(
    data: NotificationPreferenceUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """Update a notification preference"""
    service = NotificationService(db)
    pref = service.update_preference(
        user_id=current_user.id,
        notification_type=data.notification_type,
        in_app_enabled=data.in_app_enabled,
        email_enabled=data.email_enabled,
        email_digest=data.email_digest,
    )
    return NotificationPreferenceResponse.model_validate(pref)


@router.put("/preferences/bulk")
async def update_notification_preferences_bulk(
    data: BulkPreferenceUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """Update multiple notification preferences at once"""
    service = NotificationService(db)
    updated = []

    for pref_update in data.preferences:
        pref = service.update_preference(
            user_id=current_user.id,
            notification_type=pref_update.notification_type,
            in_app_enabled=pref_update.in_app_enabled,
            email_enabled=pref_update.email_enabled,
            email_digest=pref_update.email_digest,
        )
        updated.append(NotificationPreferenceResponse.model_validate(pref))

    return NotificationPreferencesResponse(preferences=updated)


@router.delete("/preferences/reset")
async def reset_notification_preferences(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """Reset all notification preferences to defaults"""
    service = NotificationService(db)
    count = service.reset_preferences(current_user.id)
    return {"reset": count}

File: frontend/src/lib/api/notifications.ts
/**
 * Notifications API
 * User notification management
 */
import { apiClient } from './client';

// ============================================================================
// Types
// ============================================================================

export interface Notification {
  id: string;
  type: string;
  title: string;
  message: string;
  priority: 'low' | 'normal' | 'high' | 'urgent';
  data?: Record<string, unknown>;
  is_read: boolean;
  read_at?: string;
  created_at: string;
}

export interface NotificationListResponse {
  notifications: Notification[];
  total: number;
  unread_count: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface NotificationCountResponse {
  unread_count: number;
}

export interface NotificationType {
  type: string;
  category: string;
  name: string;
}

export interface NotificationPreference {
  notification_type: string;
  in_app_enabled: boolean;
  email_enabled: boolean;
  email_digest: boolean;
}

export interface NotificationPreferencesResponse {
  preferences: NotificationPreference[];
}

// ============================================================================
// API Functions
// ============================================================================

/**
 * List notifications with pagination and filtering
 */
export async function listNotifications(params?: {
  page?: number;
  page_size?: number;
  unread_only?: boolean;
  notification_type?: string;
}): Promise<NotificationListResponse> {
  const searchParams = new URLSearchParams();
  if (params?.page) searchParams.set('page', params.page.toString());
  if (params?.page_size) searchParams.set('page_size', params.page_size.toString());
  if (params?.unread_only) searchParams.set('unread_only', 'true');
  if (params?.notification_type) searchParams.set('notification_type', params.notification_type);

  const query = searchParams.toString();
  const response = await apiClient.get(`/notifications${query ? `?${query}` : ''}`);
  return response.data;
}

/**
 * Get unread notification count
 */
export async function getUnreadCount(): Promise<NotificationCountResponse> {
  const response = await apiClient.get('/notifications/count');
  return response.data;
}

/**
 * Get available notification types
 */
export async function getNotificationTypes(): Promise<{ types: NotificationType[] }> {
  const response = await apiClient.get('/notifications/types');
  return response.data;
}

/**
 * Get a specific notification
 */
export async function getNotification(id: string): Promise<Notification> {
  const response = await apiClient.get(`/notifications/${id}`);
  return response.data;
}

/**
 * Mark specific notifications as read
 */
export async function markAsRead(notificationIds: string[]): Promise<{ marked_read: number }> {
  const response = await apiClient.post('/notifications/mark-read', {
    notification_ids: notificationIds,
  });
  return response.data;
}

/**
 * Mark all notifications as read
 */
export async function markAllAsRead(notificationType?: string): Promise<{ marked_read: number }> {
  const response = await apiClient.post('/notifications/mark-all-read', {
    notification_type: notificationType,
  });
  return response.data;
}

/**
 * Delete a notification
 */
export async function deleteNotification(id: string): Promise<{ deleted: boolean }> {
  const response = await apiClient.delete(`/notifications/${id}`);
  return response.data;
}

/**
 * Delete all read notifications
 */
export async function deleteAllRead(): Promise<{ deleted: number }> {
  const response = await apiClient.delete('/notifications');
  return response.data;
}

// ============================================================================
// Preferences API
// ============================================================================

/**
 * Get all notification preferences
 */
export async function getPreferences(): Promise<NotificationPreferencesResponse> {
  const response = await apiClient.get('/notifications/preferences/all');
  return response.data;
}

/**
 * Update a notification preference
 */
export async function updatePreference(
  notificationType: string,
  updates: {
    in_app_enabled?: boolean;
    email_enabled?: boolean;
    email_digest?: boolean;
  }
): Promise<NotificationPreference> {
  const response = await apiClient.put('/notifications/preferences', {
    notification_type: notificationType,
    ...updates,
  });
  return response.data;
}

/**
 * Update multiple preferences at once
 */
export async function updatePreferencesBulk(
  preferences: Array<{
    notification_type: string;
    in_app_enabled?: boolean;
    email_enabled?: boolean;
    email_digest?: boolean;
  }>
): Promise<NotificationPreferencesResponse> {
  const response = await apiClient.put('/notifications/preferences/bulk', {
    preferences,
  });
  return response.data;
}

/**
 * Reset all preferences to defaults
 */
export async function resetPreferences(): Promise<{ reset: number }> {
  const response = await apiClient.delete('/notifications/preferences/reset');
  return response.data;
}

// ============================================================================
// Notification Categories
// ============================================================================

export const NOTIFICATION_CATEGORIES = {
  system: {
    name: 'System',
    description: 'Platform announcements and maintenance notifications',
  },
  billing: {
    name: 'Billing',
    description: 'Subscription and payment notifications',
  },
  user: {
    name: 'Team',
    description: 'Team member invitations and changes',
  },
  tenant: {
    name: 'Organization',
    description: 'Organization settings and branch changes',
  },
  usage: {
    name: 'Usage',
    description: 'Quota and usage warnings',
  },
  security: {
    name: 'Security',
    description: 'Login and password notifications',
  },
};

/**
 * Get category from notification type
 */
export function getCategoryFromType(type: string): string {
  return type.split('.')[0] || 'system';
}

/**
 * Get priority color
 */
export function getPriorityColor(priority: string): string {
  switch (priority) {
    case 'urgent':
      return 'text-red-500';
    case 'high':
      return 'text-orange-500';
    case 'normal':
      return 'text-blue-500';
    case 'low':
      return 'text-gray-500';
    default:
      return 'text-gray-500';
  }
}

/**
 * Get notification icon based on type
 */
export function getNotificationIcon(type: string): string {
  const category = getCategoryFromType(type);
  switch (category) {
    case 'billing':
      return 'credit-card';
    case 'user':
      return 'users';
    case 'tenant':
      return 'building';
    case 'usage':
      return 'bar-chart';
    case 'security':
      return 'shield';
    default:
      return 'bell';
  }
}
